### ISSUE TITLE
file.copy module with recurse=true and non-existing src dir does not fail and resets dst dir permissions

### ISSUE THREAD

**Issue by sastorsl** [saltstack/salt#31356]
To reproduce:

Create a destination (dst) directory and a file in that directory.
We can see that the source directory does not exist.

```bash
$ rm -rf /tmp/test && mkdir /tmp/test && echo "hello" > /tmp/test/test.txt && ls -ld /tmp/test /tmp/test/test.txt
drwxr-xr-x 2 root root 4096 Feb 19 17:15 /tmp/test
-rw-r--r-- 1 root root    6 Feb 19 17:15 /tmp/test/test.txt
[root@ald-aot-r-salt05 ~]# ls -ld /etc/skel2
ls: cannot access /etc/skel2: No such file or directory
```

Try to copy from the non-existing directory.
I would expect this to fail - but it returns True and the directory permissions of the dst directory are reset to 000

```bash
$ sudo salt-call -l error file.copy /etc/skel2 /tmp/test/ recurse=true
local:
    True
$ ls -lda /tmp/test /tmp/test/.* /tmp/test/*
d---------   2 root root 4096 Feb 19 17:15 /tmp/test
d---------   2 root root 4096 Feb 19 17:15 /tmp/test/.
drwxrwxrwt. 15 root root 4096 Feb 19 17:15 /tmp/test/..
-rw-r--r--   1 root root    6 Feb 19 17:15 /tmp/test/test.txt
```

My version info
```bash
$ sudo salt-call --versions-report
           Salt: 2015.5.5
         Python: 2.7.5 (default, Apr  9 2015, 11:03:32)
         Jinja2: 2.7.2
       M2Crypto: 0.21.1
 msgpack-python: 0.4.6
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: Not Installed
         PyYAML: 3.10
          ioflo: Not Installed
          PyZMQ: 14.3.1
           RAET: Not Installed
            ZMQ: 3.2.5
           Mako: Not Installed
        Tornado: Not Installed
        timelib: Not Installed
       dateutil: 1.5
```

**Comment by sastorsl** [comment]
Adding os.path.exists(src) to salt/modules/file.py should fix this, but I would maybe test if recurse is set and then test with os.path.isfile(src) / os.path.isdir(src) depending on if recurse is False or True.

**Comment by jfindlay** [comment]
@sastorsl, thanks for the report.  If you think that you have an idea of how to solve this, you are welcome to send in a pull request.

**Comment by sastorsl** [comment]
Haven't contributed to github before, so I've spent the morning (europe) reading up. I'll see if I can share some inputs. I'll be AFK for at least some hours now.

**Comment by sastorsl** [comment]
Firing from the hip.
IMHO the function should fail early when something as important as the source of a copy does not exist. If you see reasons to move the test to a sub function I'll of course listen to reason.

salt/modules/file.py
```python
def copy(src, dst, recurse=False, remove_existing=False):
<snip>
    if not os.path.isabs(src):
        raise SaltInvocationError('File path must be absolute.')

    if not os.path.exists(src):
        raise CommandExecutionError('No such file or directory \'{0}.\''.format(src))
```

**Comment by sastorsl** [comment]
And an example test with a patched version of salt-2015.5.5 - salt/modules/file.py

```bash
# diff /usr/lib/python2.7/site-packages/salt/modules/file.py{_20160220,}
2398a2399,2401
>     if not os.path.exists(src):
>         raise CommandExecutionError('No such file or directory \'{0}.\''.format(src))
>
```

Performing some tests
```bash
   # A file not existing
# salt-call -l error file.copy /etc/passwds /tmp/test/test.txt
Error running 'file.copy': No such file or directory '/etc/passwds.'

  # A directory not existing
# salt-call -l error file.copy /etc/skel_does_not_exist /tmp/test/ recurse=true
Error running 'file.copy': No such file or directory '/etc/skel_does_not_exist.'

   # A file which exists
# salt-call -l error file.copy /etc/passwd /tmp/test/test.txt
local:
    True

  # A directory which exists
# salt-call -l error file.copy /etc/skel /tmp/test/ recurse=true
local:
    True
```

**Comment by jfindlay** [comment]
@sastorsl, this is good.  You are welcome to ping me if you have trouble creating a pull request.
